gapic/src/main/com/google/gapid/views/DeviceDialog.java

```diff
@@ -0,0 +1,334 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.gapid.views;
+
+import static com.google.gapid.util.Logging.throttleLogRpcError;
+import static com.google.gapid.util.MoreFutures.logFailure;
+import static com.google.gapid.widgets.Widgets.createDropDownViewer;
+import static com.google.gapid.widgets.Widgets.createGroup;
+import static com.google.gapid.widgets.Widgets.createLabel;
+import static com.google.gapid.widgets.Widgets.createLink;
+import static com.google.gapid.widgets.Widgets.scheduleIfNotDisposed;
+import static com.google.gapid.widgets.Widgets.withIndents;
+import static com.google.gapid.widgets.Widgets.withLayoutData;
+import static java.util.logging.Level.WARNING;
+
+import com.google.gapid.models.Capture;
+import com.google.gapid.models.Devices;
+import com.google.gapid.models.Devices.DeviceCaptureInfo;
+import com.google.gapid.models.Devices.DeviceValidationResult;
+import com.google.gapid.models.Models;
+import com.google.gapid.proto.device.Device;
+import com.google.gapid.rpc.Rpc;
+import com.google.gapid.rpc.RpcException;
+import com.google.gapid.rpc.SingleInFlight;
+import com.google.gapid.rpc.UiErrorCallback;
+import com.google.gapid.util.Loadable;
+import com.google.gapid.util.Messages;
+import com.google.gapid.util.Paths;
+import com.google.gapid.util.Scheduler;
+import com.google.gapid.util.URLs;
+import com.google.gapid.widgets.DialogBase;
+import com.google.gapid.widgets.LoadingIndicator;
+import com.google.gapid.widgets.Widgets;
+
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.jface.viewers.ArrayContentProvider;
+import org.eclipse.jface.viewers.ComboViewer;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.jface.viewers.LabelProvider;
+import org.eclipse.jface.viewers.StructuredSelection;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.layout.GridLayout;
+import org.eclipse.swt.program.Program;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Group;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Link;
+import org.eclipse.swt.widgets.Shell;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+/**
+ * View responsible to show a replay device selection dialog when need be.
+ */
+public class DeviceDialog implements Devices.Listener, Capture.Listener {
+  protected static final Logger LOG = Logger.getLogger(DeviceDialog.class.getName());
+
+  private final Shell shell;
```

Comment:
No need for this, just use `parent.getShell()` to get the shell.

Reply:
done